exercises/tom_clayton/ex02_oo_basics/src/ShapeSorter.cpp

```diff
@@ -0,0 +1,111 @@
+#include "ShapeSorter.h"
+#include <iostream>
+#include <algorithm>
+#include <vector>
+
+/** Comparator function to return true if s1 area > s2 area
+ *  @param s1 reference to the first shape
+ *  @param s2 reference to the second shape
+ * 
+ *  @return boolean to determine larger area
+ */ 
+static bool areaCompare(std::shared_ptr<Shape>& s1, std::shared_ptr<Shape>& s2) {
+  return (s1->getArea() > s2->getArea());
+}
+
+/** Comparator function to return true if s1 perimeter > s2 perimeter
+ *  @param s1 reference to the first shape
+ *  @param s2 reference to the second shape
+ * 
+ *  @return boolean to determine larger perimeter
+ */ 
+static bool periCompare(std::shared_ptr<Shape> s1, std::shared_ptr<Shape> s2) {
+  return (s1->getPerimeter() > s2->getPerimeter());
+}
+
+//ShapeSorter constructor
+ShapeSorter::ShapeSorter(std::vector<std::shared_ptr<Shape>> shpVect) {
+  m_shpVect = shpVect;
+  m_vecLen = shpVect.size();
+}
+
+/** Print out all shapes of a given type, or an error message if no
+ *  shapes of that type are found
+ *  @param string type for type of string
+ */ 
+const void ShapeSorter::printShapesType(const std::string type) {
+  bool shapeTypeFound = false;
+  for(auto shape_ptr : m_shpVect) {

```

Comment:
You can use const references here as we are not changing each shape within the loop
 i.e `for(const &auto shape_ptr : m_shpVect) `

Reply:
Similarly for the other for loops in this file